common/src/main/java/org/figuramc/figura/model/FiguraModelPart.java

```diff
@@ -89,15 +93,20 @@ public FiguraModelPart(Avatar owner, String name, @Nullable String uuid, PartCus
         this.children = children;
         this.collectionInfo = collectionInfo;
         if (collections != null) {
-            for (String n: collections) this.collections.put(n, new ArrayList<>());
-            walkCollections(this.collections, collections);
+            this.collections = new HashMap();
+            var builder = new HashMap();
+            for (String n: collections) builder.put(n, new ImmutableSet.Builder());
+            walkCollections(builder, collections);
+            builder.forEach((BiConsumer<String, ImmutableSet.Builder>) (key, parts) -> this.collections.put(key, new PartCollection(() -> owner.luaRuntime.typeManager, parts.build())));
+        } else {
+            this.collections = null;
         }
     }
 
-    private void walkCollections(Map<String, List<FiguraModelPart>> root, String names[]) {
+    private void walkCollections(Map<String, ImmutableSet.Builder> root, String names[]) {
```

Comment:
`ImmutableSet.Builder` has a type parameter, and it should be `String[] names` instead of `String names[]` (convention, not requirement)